packages/protocol/engine/consensus/acceptance/gadget.go

```diff
@@ -158,29 +221,40 @@ func (a *Gadget) block(id models.BlockID) (block *Block, exists bool) {
 	return storage.Get(id)
 }
 
-func (a *Gadget) propagateAcceptance(marker markers.Marker) (queuedBlocks []*Block) {
+func (a *Gadget) propagateAcceptance(marker markers.Marker, confirmed bool) (acceptedBlocks, confirmedBlocks []*Block) {
 	bookerBlock, blockExists := a.tangle.BlockFromMarker(marker)
 	if !blockExists {
 		return
 	}
 
 	block, blockExists := a.GetOrRegisterBlock(bookerBlock.ID())
-	if !blockExists || block.IsAccepted() {
+	if !blockExists || block.IsAccepted() && !confirmed || block.IsConfirmed() && confirmed {
 		// this can happen when block was a root block and while processing this method, the root blocks method has already been replaced
 		return
 	}
 
 	pastConeWalker := walker.New[*Block](false).Push(block)
 	for pastConeWalker.HasNext() {
 		walkerBlock := pastConeWalker.Next()
-		if !walkerBlock.SetQueued() {
-			continue
+
```

Comment:
We don't want to walk twice, that's why the confirmation is huddled together with the acceptance, right? But this makes the code a bit ugly and two distinct concepts are mixed together. Maybe we can rethink this a little bit so that the current `Gadget` becomes a wrapper component that houses the blocks, makes the walk etc but each actual `Gadget` for `acceptance` and `confirmation` lives in a separate struct. Each owning their own causal order, last*Marker, etc. These components seem so similar, maybe we can even make one implementation and make it configurable. Wdyt?

Reply:
> We don't want to walk twice, that's why the confirmation is huddled together with the acceptance 

This, and also because acceptance threshold depends on whether there is enough online weight to confirm. That makes the two concepts intertwined, as there is not much code that is unique to either confirmation or acceptance. It seems like separating those two concepts while retaining single walk would complicate the code.